src/analyzer/analyzersilence.cpp

```diff
@@ -117,29 +102,37 @@ void AnalyzerSilence::storeResults(TrackPointer tio) {
     double introStart = mixxx::kAnalysisChannels * m_iSignalStart;
     double outroEnd = mixxx::kAnalysisChannels * m_iSignalEnd;
 
-    if (shouldUpdateMainCue(tio->getCuePoint())) {
-        tio->setCuePoint(CuePosition(introStart, Cue::AUTOMATIC));
+    CuePointer pAudibleSound = pTrack->findCueByType(Cue::Type::AudibleSound);
+    if (pAudibleSound == nullptr || pAudibleSound->getLength() <= 0) {
+        pAudibleSound = pTrack->createAndAddCue();
+        pAudibleSound->setType(Cue::Type::AudibleSound);
+        pAudibleSound->setPosition(introStart);
+        pAudibleSound->setLength(outroEnd - introStart);
     }
 
-    CuePointer pIntroCue = tio->findCueByType(Cue::INTRO);
-    if (!pIntroCue) {
-        pIntroCue = tio->createAndAddCue();
-        pIntroCue->setType(Cue::INTRO);
-        pIntroCue->setSource(Cue::AUTOMATIC);
+    CuePointer pIntroCue = pTrack->findCueByType(Cue::Type::Intro);
+
+    double oldMainCue = pTrack->getCuePoint().getPosition();
+    // NOTE: the actual default for this ConfigValue is set in DlgPrefDeck.
+    if (m_pConfig->getValue(ConfigKey("[Controls]", "SetIntroStartAtMainCue"), false)
+            && pIntroCue == nullptr && oldMainCue != kCueNotSet && oldMainCue != 0.0) {
+        introStart = oldMainCue;
+    } else if (oldMainCue == kCueNotSet) {
+        pTrack->setCuePoint(CuePosition(introStart));
     }
-    if (pIntroCue->getSource() != Cue::MANUAL) {
+
+    if (!pIntroCue) {
+        pIntroCue = pTrack->createAndAddCue();
+        pIntroCue->setType(Cue::Type::Intro);
         pIntroCue->setPosition(introStart);
         pIntroCue->setLength(0.0);
     }
 
-    CuePointer pOutroCue = tio->findCueByType(Cue::OUTRO);
+    CuePointer pOutroCue = pTrack->findCueByType(Cue::Type::Outro);
     if (!pOutroCue) {
-        pOutroCue = tio->createAndAddCue();
-        pOutroCue->setType(Cue::OUTRO);
-        pOutroCue->setSource(Cue::AUTOMATIC);
-    }
-    if (pOutroCue->getSource() != Cue::MANUAL) {
-        pOutroCue->setPosition(-1.0);
+        pOutroCue = pTrack->createAndAddCue();
+        pOutroCue->setType(Cue::Type::Outro);
+        pOutroCue->setPosition(kCueNotSet);
         pOutroCue->setLength(outroEnd);
```

Comment:
The length is used as the end point if no start point is set. This is an ugly consequence of storing ranges as a start and a length rather than a pair of points. See Cue::getEndPosition in src/track/cue.cpp

Reply:
Ah, ok. Please add a source code comment.

Reply:
I have instead added a Cue::setEndPoint method that is easier to work with and used it here in AnalyzerSilence. In the future, this could replace the set/getLength methods and the cue length could just be an implementation detail. But there is no need to refactor existing code using get/setLength in this PR.